iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp

```diff
@@ -24,6 +24,67 @@
 namespace mlir {
 namespace iree_compiler {
 
+void ConvertToDynamicSharedMemory(ModuleOp moduleOp) {
+  // Collect all the adressOfOps to static shared memory globals.
+  SmallVector<LLVM::AddressOfOp> addressOfOps;
+  moduleOp.walk([&](LLVM::AddressOfOp addressOfOp) {
+    if (!addressOfOp.getGlobal().getConstant())
+      addressOfOps.push_back(addressOfOp);
+  });
+  if (addressOfOps.size()) {
+    OpBuilder builder(moduleOp);
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPoint(&moduleOp.front());
+    LLVM::GlobalOp global;  // new shared memory global
+    LLVM::AddressOfOp globalPtr;
+    auto type =
+        LLVM::LLVMArrayType::get(IntegerType::get(builder.getContext(), 8), 0);
+    global = builder.create<LLVM::GlobalOp>(
+        moduleOp.getLoc(), type, /*isConstant=*/false, LLVM::Linkage::External,
+        "__dynamic_shared_memory__", Attribute(),
+        /*alignment=*/4, /*addr_space=*/3);
+    int numberOfBytes = 0;
+    // Replace the addressOfOps with correctly offseted pointers to dynamic
+    // shared memory
+    for (auto addressOfOpsIt : llvm::enumerate(addressOfOps)) {
+      auto addressOfOp = addressOfOpsIt.value();
+      auto loc = addressOfOp.getLoc();
+      builder.setInsertionPoint(addressOfOp);
+      auto globalOp = addressOfOp.getGlobal();
+      globalPtr = builder.create<LLVM::AddressOfOp>(loc, global);
+      Value zero = builder.createOrFold<LLVM::ConstantOp>(
+          loc, IntegerType::get(builder.getContext(), 64),
+          builder.getI64IntegerAttr(0));
+      Value offset = builder.createOrFold<LLVM::ConstantOp>(
+          loc, IntegerType::get(builder.getContext(), 64),
+          builder.getI64IntegerAttr(numberOfBytes));
+      Value shiftedPtr = builder.createOrFold<LLVM::GEPOp>(
+          loc, globalPtr.getType(), globalPtr, ValueRange({zero, offset}));
+      Value castPtr = builder.create<LLVM::BitcastOp>(
+          loc,
+          LLVM::LLVMPointerType::get(globalOp.getType(), global.getAddrSpace()),
+          shiftedPtr);
+      addressOfOp.replaceAllUsesWith(castPtr);
+      auto thisarray = globalOp.getType().dyn_cast<LLVM::LLVMArrayType>();
+      // calculate offset for next allocation
+      DataLayout dataLayout = DataLayout::closest(addressOfOp);
+      numberOfBytes += dataLayout.getTypeSizeInBits(thisarray) / 8;
+      globalOp.erase();
```

Comment:
This assumes that globalOp has only one use. Even though it is most likely cse would merge all the addressOf this is not completely safe. Here I'm surprised this doesn't assert as you are erasing the globalOp while it still has a use?